File: #5 Tetris_Game_Abstract_Implementation/tetromino.cpp
#include <iostream>
#include <random>
#include "tetromino.hpp"
using namespace std;

    int Tetromino::numberofmoves=0;

    Tetromino::Tetromino(tetris t):tet(t)//Tetromino one parameter constructor to take tetris enum ...
    {
        create_tetrominos();//create tetromino after called tetromino obj...
        numberofmoves++;
    }

    Tetromino::~Tetromino()//delete tetromino obj after finish its job...
    {   
        // for(int i=0;i<tet_row;i++)
        //     delete[] tetromino[i];
        // delete [] tetromino;
    }
    const Tetromino &Tetromino::operator=(const Tetromino &oth)
    {
        cout<<"operatpor= worked!"<<endl;
        tet_row=oth.getTetRow();
        tet_col=oth.getTetCol();

        tetromino=new char *[tet_row];
        for(int i=0;i<tet_row;i++)
            tetromino[i]=new char[tet_col];

        for(int i=0;i<tet_row;i++)
            for(int j=0;j<tet_col;j++)
                tetromino[i][j]=oth.tetromino[i][j];
        
        return *this;
    }

    void Tetromino::print_tetromino()const{//print tetrominoon the screen
        for(int i=0;i<tet_row;i++){
            for(int j=0;j<tet_col;j++){
                cout<<tetromino[i][j]; 
            }
            cout<<endl;
        }      
    }

    void Tetromino::create_arr(int row,int col){//create tetromino obj array dynamicaly ...
        tet_row=row;
        tet_col=col;

        tetromino=new char*[row];
        for(int i=0;i<row;i++)
            tetromino[i]=new char[col];

        for(int i=0;i<row;i++)
            for(int j=0;j<col;j++)
                tetromino[i][j]=' '; 
    }

    void Tetromino::rotatematrix(char direction){//rotate matrix left or right..
        int i,j;

        int k=0,m=0;

        char **temp=new char*[tet_col];//create dynamic temp array to hold rotated object...
        for(int i=0;i<tet_col;i++)
            temp[i]=new char[tet_row];


    	if(direction=='R'){ //Turn Right
            for (i = 0; i < tet_col; i++) {
    		    for (j = tet_row-1; j >= 0; j--) {
                    temp[k][m]=tetromino[j][i];
                    m++;
    		    }
                k++;
                m=0;
            }
    	}
        else{
            for (i = tet_col-1; i>=0; i--) {//Turn Left
    	        for (j = 0; j < tet_row; j++) {
                    temp[k][m]=tetromino[j][i];
                    m++;
    	        }
                k++;
                m=0;
    	    }
        }

        for(int i=0;i<tet_row;i++)//delete old tetromino object...
            delete[] tetromino[i];

        delete [] tetromino;

        tetromino=new char*[tet_col];//create new tetromino object with row and column of rotated object
        for(int i=0;i<tet_col;i++)
            tetromino[i]=new char[tet_row];


        for(int i=0;i<tet_col;i++) //tetromino=temp;
            for(int j=0;j<tet_row;j++)
                tetromino[i][j]=temp[i][j];


        for(int i=0;i<tet_col;i++)//delete temp object after assign tetromino...
            delete[] temp[i];

        delete [] temp;

        int temp1=tet_row;
        tet_row=tet_col;
        tet_col=temp1;

        numberofmoves++;
    }

    void Tetromino::create_tetrominos(){
        switch (getTetro())//Creates objects according to the enum parameter sent...
        {
            case I:
                create_arr(1,4);
                tetromino[0][0]='I',tetromino[0][1]='I',tetromino[0][2]='I',tetromino[0][3]='I';
                break;
            case O:
                create_arr(2,2);
                tetromino[0][0]='O',tetromino[0][1]='O',tetromino[1][0]='O',tetromino[1][1]='O';
                break;
            case T:
                create_arr(2,3);
                tetromino[0][0]='T',tetromino[0][1]='T',tetromino[0][2]='T',tetromino[1][1]='T';
                break;
            case J:
                create_arr(3,2);
                tetromino[0][1]='J',tetromino[1][1]='J',tetromino[2][0]='J',tetromino[2][1]='J';
                break;
            case L:
                create_arr(3,2);
                tetromino[0][0]='L',tetromino[1][0]='L',tetromino[2][0]='L',tetromino[2][1]='L';
                break;

            case S:
                create_arr(2,3);
                tetromino[0][1]='S',tetromino[0][2]='S',tetromino[1][0]='S',tetromino[1][1]='S';
                break;
            case Z:
                create_arr(2,3);
                tetromino[0][0]='Z',tetromino[0][1]='Z',tetromino[1][1]='Z',tetromino[1][2]='Z';
                break;
        }
    }

File: #7 Tetris_Game_JAVA/Tetromino.java

public class Tetromino {

    private char objType;
    private int _R;
    private int _C;
    private char [][] tetShape;

    public Tetromino()//no param const
    {
        //intentionaly empty
    }

    public Tetromino(char T)//one param const
    {
        objType=T;
        create_Tetrominos();
    }

    public void setType(char type)
    {
        objType=type;
    }

    public char getType()
    {
        return objType;
    }

    public int get_R()
    {
        return _R;
    }


    public int get_C()
    {
        return _C;
    }

    public char[][] getShape()
    {
        return tetShape;
    }

    private void create_shape(int row,int col)//create tetromino obj given row and col size...
    {
        _R=row;
        _C=col;

        tetShape = new char[_R][_C];
        
        for(int i=0;i<row;i++)
            for(int j=0;j<col;j++)
                tetShape[i][j]=' ';
    } 

    public void rotate_shape(char direction)//rotate tetromino object given direction
    {
        int i,j,k=0,m=0;

        char [][] temp=new char[_C][_R];

    	if(direction=='R') //Turn Right
        { 
            for (i = 0; i < _C; i++)
            {
    		    for (j = _R-1; j >= 0; j--)
                {
                    temp[k][m]=tetShape[j][i];
                    m++;
    		    }
                k++;
                m=0;
            }
    	}
        else{
            for (i = _C-1; i>=0; i--) //Turn Left
            {
    	        for (j = 0; j < _R; j++) 
                {
                    temp[k][m]=tetShape[j][i];
                    m++;
    	        }
                k++;
                m=0;
    	    }
        }

        tetShape=null;
        tetShape=new char[_C][_R];

        for(i=0;i<_C;i++) //tetromino=temp;
            for(j=0;j<_R;j++)
                tetShape[i][j]=temp[i][j];

        temp=null;

        int temp1=_R;
        _R=_C;
        _C=temp1;
    }
    
    public void printShape()
    {
        for(int i=0;i<_R;i++)
        {
            for(int j=0;j<_C;j++)
            {
                System.out.print(tetShape[i][j]); 
            }
            System.out.printf("\n");
        }  
    }

    public void create_Tetrominos()
    {
        switch (getType())//Creates objects according to the enum parameter sent...
        {
            case 'I':
                create_shape(1,4);
                tetShape[0][0]='I';
                tetShape[0][1]='I';
                tetShape[0][2]='I';
                tetShape[0][3]='I';
                break;
            case 'O':
                create_shape(2,2);
                tetShape[0][0]='O';
                tetShape[0][1]='O';
                tetShape[1][0]='O';
                tetShape[1][1]='O';
                break;
            case 'T':
                create_shape(2,3);
                tetShape[0][0]='T';
                tetShape[0][1]='T';
                tetShape[0][2]='T';
                tetShape[1][1]='T';
                break;
            case 'J':
                create_shape(3,2);
                tetShape[0][1]='J';
                tetShape[1][1]='J';
                tetShape[2][0]='J';
                tetShape[2][1]='J';
                break;
            case 'L':
                create_shape(3,2);
                tetShape[0][0]='L';
                tetShape[1][0]='L';
                tetShape[2][0]='L';
                tetShape[2][1]='L';
                break;

            case 'S':
                create_shape(2,3);
                tetShape[0][1]='S';
                tetShape[0][2]='S';
                tetShape[1][0]='S';
                tetShape[1][1]='S';
                break;
            case 'Z':
                create_shape(2,3);
                tetShape[0][0]='Z';
                tetShape[0][1]='Z';
                tetShape[1][1]='Z';
                tetShape[1][2]='Z';
                break;
        }
    }
}
